doc: Add ubuntu-image security overview (#2102)
canonical/subiquity: doc/explanation/ubuntu-image-security-overview.rst

.. _ubuntu-image-security-overview:

ubuntu-image security overview
==============================

Overview of security aspects of ubuntu-image.

Privileged execution
--------------------
 

> ``ubuntu-image`` requires **elevated permissions** to properly run. It is recommended to use a dedicated building machine, make sure ``ubuntu-image`` is installed from a trusted source and the provided configuration is trusted. 

Reviewer: ```suggestion
``ubuntu-image`` requires **elevated permissions** to properly run. It is recommended to use a dedicated building machine. Make sure ``ubuntu-image`` is installed from a trusted source, and the provided configuration is trusted. 
```

Cryptography
------------

> ``ubuntu-image`` is a wrapper around several lower level tools to build a Ubuntu image. Use of cryptographic technologies can then be grouped in two categories: direct usage in the tool or in one of the wrapped tools.

Reviewer: ```suggestion
``ubuntu-image`` is a wrapper around several lower level tools to build an Ubuntu image. Therefore, the use of cryptographic technologies can be divided into two categories:

* Direct use in the ``ubuntu-image`` tool
* Indirect use in one of the wrapped tools
```

> ubuntu-image itself and Go libraries
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Cryptography in ubuntu-image itself and Go libraries
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> * Tarball SHA256 checksum: When building a ``classic`` image, user can provide a tarball serving as the rootfs content of the image. The SHA256 checksum of the tarball can optionally be provided to verify it has not been altered. This verification `crypto/sha256`_ Go standard library package, implementing the SHA256 hash algorithm as defined in FIPS 180-4.

Reviewer: ```suggestion
* Tarball SHA256 checksum: When building a ``classic`` image, user can provide a tarball serving as the rootfs content of the image. The SHA256 checksum of the tarball can optionally be provided to verify it has not been altered. This verification uses the `crypto/sha256`_ Go standard library package, which implements the SHA256 hash algorithm as defined in FIPS 180-4.
```

> * Disk UUIDs: To generate unique disk IDs, ``ubuntu-image`` uses the ``Read()`` function from the `crypto/rand`_ Go standard library package, implementing a cryptographically secure random number generator.

Reviewer: ```suggestion
* Disk UUIDs: To generate unique disk IDs, ``ubuntu-image`` uses the ``Read()`` function from the `crypto/rand`_ Go standard library package, which implements a cryptographically secure random number generator.
```

> * Model assertion signature: When building a ``snap`` image, user must provide a signed model assertion. Verification of this signature is handled by the `snapd library`_, relying on SHA3-384 and SHA512 for hashing and OpenPGP v4 signatures with RSA 4096/8192 keys.

Reviewer: ```suggestion
* Model assertion signature: When building a ``snap`` image, the user must provide a signed model assertion. Verification of this signature is handled by the `snapd library`_,  which relies on SHA3-384 and SHA512 for hashing and on OpenPGP v4 signatures with RSA 4096/8192 keys.
```

> Underlying tools
> ~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Cryptography in wrapped tools
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> * PPA handling: During the build, PPAs can be added to the ``apt`` configuration of the image. The fingerprint of the PPA can be declared in the configuration provided to ``ubuntu-image``. This value is passed down to ``gpg`` to fetch the keys to validate the packages signature. The correct handling of these keys is delegated to ``gpg`` up to the point where ``ubuntu-image`` write them to the resulting ``apt`` configuration.

Reviewer: ```suggestion
* PPA handling: During the build, PPAs (Personal Package Archives) can be added to the ``apt`` configuration of the image. The fingerprint of the PPA can be declared in the configuration provided to ``ubuntu-image``. This value is passed to ``gpg`` to fetch the keys to validate the package signatures. The correct handling of these keys is delegated to ``gpg`` up to the point where ``ubuntu-image`` writes them to the resulting ``apt`` configuration.
```

Author: I am not sure about the suggestion to replace `packages signature` with `package signature`. I was talking about the signature of each package, not the multiple signatures of a single package.

> * Downloaded resources: During the build various resources are fetched from remote sources (seeds, packages, gadget trees). ``ubuntu-image`` is not enforcing the usage of encrypted communication channels but if they are used (HTTPS, git over SSH, etc.) their correct handling is delegated to the tool actually pulling the resource (germinate, git, etc.). In this case, if any security-related error is making a tool fail, ubuntu-image will fail too and will display the error so the user can be alerted and take any appropriate measure to remediate. 

Reviewer: ```suggestion
* Downloaded resources: During the build, various resources are fetched from remote sources (seeds, packages, gadget trees). ``ubuntu-image`` is not enforcing the use of encrypted communication channels, but if they are used (HTTPS, Git over SSH, etc.), their correct handling is delegated to the tool actually pulling the resource (Germinate, Git, etc.). In this case, if any security-related error is causing a tool to fail, ``ubuntu-image`` also fails and displays the error, so the user is alerted and can take appropriate measures to remedy the problem.
```

> Misc.
> ~~~~~

Reviewer: ```suggestion
Miscellaneous
~~~~~~~~~~~~~
```

* Secrets (passwords and hashes) can be present in the configuration files (image definition YAML) provided to ``ubuntu-image`` to build images. Specifically:

>   * In the "manual" customization section, user accounts can be defined with plain text or hashed passwords. These values are directly passed down to ``chpassword`` without any treatment.

Reviewer: ```suggestion
  * In the ``manual`` customisation section, user accounts can be defined with plain text or hashed passwords. These values are directly passed to the ``chpassword`` utility without any treatment.
```

>   * In the "cloud-init" customization section, given cloud-init configuration can contain hashed password. These values are written in `cloud-init`_ configuration files without any treatment.

Reviewer: ```suggestion
  * In the ``cloud-init`` customisation section, the given cloud-init configuration can contain hashed passwords. These values are written into `cloud-init`_ configuration files without any treatment.
```

>   These configuration files should then be securely store and if secrets are used they should ideally be injected at runtime.

Reviewer: ```suggestion
  These configuration files should then be securely stored, and if secrets are used, they should ideally be injected at runtime.
```

Cryptography in wrapped tools
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* PPA handling: During the build, PPAs (Personal Package Archives) can be added to the ``apt`` configuration of the image. The fingerprint of the PPA can be declared in the configuration provided to ``ubuntu-image``. This value is passed to ``gpg`` to fetch the keys to validate the package signatures. The correct handling of these keys is delegated to ``gpg`` up to the point where ``ubuntu-image`` writes them to the resulting ``apt`` configuration.

* Downloaded resources: During the build, various resources are fetched from remote sources (seeds, packages, gadget trees). ``ubuntu-image`` is not enforcing the use of encrypted communication channels, but if they are used (HTTPS, Git over SSH, etc.), their correct handling is delegated to the tool actually pulling the resource (Germinate, Git, etc.). In this case, if any security-related error is causing a tool to fail, ``ubuntu-image`` also fails and displays the error, so the user is alerted and can take appropriate measures to remedy the problem.

Miscellaneous
~~~~~~~~~~~~~

> * Secrets (passwords and hashes) can be present in the configuration files (image definition YAML) provided to ``ubuntu-image`` to build images. Specifically:

Reviewer 2: Should we also include information about private PPA support? Since when a private PPA is added, the yaml needs to have the token included in as well (user:auth).
